File: modules/brl/native/fileinputstream.java

class BBFileInputStream extends BBStream{

	boolean Open( String path ){
		if( _stream!=null ) return false;
						
		try{
			_stream=BBGame.Game().OpenInputStream(path);
			if( _stream!=null ){
				_position=0;
				_length=_stream.available();
				return true;
			}
		
		}catch( IOException ex ){
			throw new RuntimeException(ex);
		}
		
		_stream=null;
		_position=0;
		_length=0;
		return false;
	}

	void Close(){
		if( _stream==null ) return;

		try{
			_stream.close();
		}catch( IOException ex ){
		}
		_stream=null;
		_position=0;
		_length=0;
	}
	
	int Eof(){
		if( _stream==null ) return -1;
		
		return (_position==_length) ? 1 : 0;
	}
	
	int Length(){
		return (int)_length;
	}
	
	int Offset(){
		return (int)_position;
	}
	
	int Seek( int offset ){
		return 0;
	}
		
	int Read( BBDataBuffer buffer,int offset,int count ){
		if( _stream==null ) return 0;
		
		try{
			int n=_stream.read( buffer._data.array(),offset,count );
			if( n>=0 ){
				_position+=n;
				return n;
			}
		}catch( IOException ex ){
		}
		return 0;
	}
	
	int Write( BBDataBuffer buffer,int offset,int count ){
		return 0;
	}

	InputStream _stream;
	long _position,_length;
}

File: modules/brl/native/fileinputstream.cpp

// ***** filestream.h *****

class BBFileInputStream : public BBStream{
public:

	BBFileInputStream();
	~BBFileInputStream();

	void Close();
	int Eof();
	int Length();
	int Position();
	int Seek( int position );
	int Read( BBDataBuffer *buffer,int offset,int count );
	int Write( BBDataBuffer *buffer,int offset,int count );

	bool Open( String path,String mode );
	
private:
	FILE *_file;
	int _position;
	int _length;
};

// ***** fileinputstream.cpp *****

BBFileInputStream::BBFileInputStream():_file(0),_position(0),_length(0){
}

BBFileInputStream::~BBFileInputStream(){
	if( _file ) fclose( _file );
}

bool BBFileInputStream::Open( String path,String mode="r" ){
	if( _file ) return false;

	String fmode;	
	if( mode=="r" ){
		fmode="rb";
	}else if( mode=="w" ){
		fmode="wb";
	}else if( mode=="u" ){
		fmode="rb+";
	}else{
		return false;
	}

	_file=BBGame::Game()->OpenFile( path,fmode );
	if( !_file && mode=="u" ) _file=BBGame::Game()->OpenFile( path,"wb+" );
	if( !_file ) return false;
	
	fseek( _file,0,SEEK_END );
	_length=ftell( _file );
	fseek( _file,0,SEEK_SET );
	_position=0;
	
	return true;
}

void BBFileInputStream::Close(){
	if( !_file ) return;
	
	fclose( _file );
	_file=0;
	_position=0;
	_length=0;
}

int BBFileInputStream::Eof(){
	if( !_file ) return -1;
	
	return _position==_length;
}

int BBFileInputStream::Length(){
	return _length;
}

int BBFileInputStream::Position(){
	return _position;
}

int BBFileInputStream::Seek( int position ){
	if( !_file ) return 0;
	
	fseek( _file,position,SEEK_SET );
	_position=ftell( _file );
	return _position;
}

int BBFileInputStream::Read( BBDataBuffer *buffer,int offset,int count ){
	if( !_file ) return 0;
	
	int n=fread( buffer->WritePointer(offset),1,count,_file );
	_position+=n;
	return n;
}

int BBFileInputStream::Write( BBDataBuffer *buffer,int offset,int count ){
	if( !_file ) return 0;
	
	int n=fwrite( buffer->ReadPointer(offset),1,count,_file );
	_position+=n;
	if( _position>_length ) _length=_position;
	return n;
}
